providers.bzl

```diff
@@ -0,0 +1,129 @@
+"""Providers for interop between JS rules.
+
+This file has to live in the built-in so that all rules can load() the providers
+even if users haven't installed any of the packages/*
+
+These providers allows rules to interoperate without knowledge
+of each other.
+
+You can think of a provider as a message bus.
+A rule "publishes" a message as an instance of the provider, and some other rule
+subscribes to these by having a (possibly transitive) dependency on the publisher.
+
+## Debugging
+
+Debug output is considered orthogonal to these providers.
+Any output may or may not have user debugging affordances provided, such as
+readable minification.
+We expect that rules will have a boolean `debug` attribute, and/or accept the `DEBUG`
+environment variable.
+Note that this means a given build either produces debug or non-debug output.
+If users really need to produce both in a single build, they'll need two rules with
+differing 'debug' attributes.
+"""
+
+JSModuleInfo = provider(
+    doc = """TODO""",
+    fields = {
+        "module_format": "a string like [amd, cjs, esm, iife, umd]",
+        "sources": "depset of JavaScript files",
```

Comment:
Where do sourcemaps live? I thought there would be a `sourcemaps` property beside this (not sure where I got that thought though 🤔).

Reply:
Thinking the would live together now as they always want to travel together except in a few leaf cases such as bundling when you want to filter them out. Having them separated in the provider was leading to a lot of extra boilerplate in ever rule & aspect the provider was used. Trying this out for size.

Reply:
That sounds good, we should update this field description though.